File: riceCooker_test.rb
require 'minitest/autorun'
require_relative 'riceCooker'

class RiceCookerTest < Minitest::Test
  def setup
    @console_spy = []
    def @console_spy.log(msg)
      self << msg
    end

    def @console_spy.restore
      @spied_obj = nil
    end
  end

  def set_rice_and_water_added(value)
    RiceCooker.instance_variable_set(:@rice_and_water_added, value)
  end

  def test_initial_state
    assert_equal RiceCooker::STATE_OFF, RiceCooker.instance_variable_get(:@rice_cooker_state)
    refute RiceCooker.instance_variable_get(:@rice_and_water_added)
  end

  def test_display_state
    RiceCooker.display_state
    assert_includes @console_spy, "\nCurrent state of the rice cooker: #{RiceCooker::STATE_OFF}"
  end

  def test_plug_in_without_rice_and_water
    set_rice_and_water_added(false)
    RiceCooker.plug_in
    assert_includes @console_spy, 'Error: Add rice and water before plugging in and starting cooking.'
    refute RiceCooker.instance_variable_get(:@rice_cooker_state) == RiceCooker::STATE_COOKING
  end

  def test_plug_in
    set_rice_and_water_added(true)
    RiceCooker.plug_in
    assert_includes @console_spy, 'The rice cooker is plugged in, and cooking begins.'
    assert_equal RiceCooker::STATE_COOKING, RiceCooker.instance_variable_get(:@rice_cooker_state)
  end

  def test_plug_in_already_cooking
    set_rice_and_water_added(true)
    RiceCooker.instance_variable_set(:@rice_cooker_state, RiceCooker::STATE_COOKING)
    RiceCooker.plug_in
    assert_includes @console_spy, 'Error: The rice cooker is already cooking or in keep warm mode.'
    assert_equal RiceCooker::STATE_COOKING, RiceCooker.instance_variable_get(:@rice_cooker_state)
  end

  def test_plug_in_already_keep_warm
    set_rice_and_water_added(true)
    RiceCooker.instance_variable_set(:@rice_cooker_state, RiceCooker::STATE_KEEP_WARM)
    RiceCooker.plug_in
    assert_includes @console_spy, 'Error: The rice cooker is already cooking or in keep warm mode.'
    assert_equal RiceCooker::STATE_KEEP_WARM, RiceCooker.instance_variable_get(:@rice_cooker_state)
  end

  def test_finish_cooking
    set_rice_and_water_added(true)
    RiceCooker.instance_variable_set(:@rice_cooker_state, RiceCooker::STATE_COOKING)
    RiceCooker.finish_cooking
    assert_includes @console_spy, 'Cooking is finished. The rice cooker is in keep warm mode.'
    assert_equal RiceCooker::STATE_KEEP_WARM, RiceCooker.instance_variable_get(:@rice_cooker_state)
  end

  def test_finish_cooking_no_cooking_in_progress
    set_rice_and_water_added(true)
    RiceCooker.finish_cooking
    assert_includes @console_spy, 'Error: No cooking in progress.'
    assert_equal RiceCooker::STATE_OFF, RiceCooker.instance_variable_get(:@rice_cooker_state)
  end

  def test_quit_program
    RiceCooker.quit_program
    assert_includes @console_spy, "\n======================================================================="
    assert_includes @console_spy, 'Goodbye! Thanks for using the rice cooker program.'
  end

  def teardown
    @console_spy.restore
  end
end

File: riceCooker_test.go
package main

import (
	"fmt"
	"testing"
)

type testLogger struct {
	consoleOutput string
}

func (l *testLogger) Log(a ...interface{}) {
	l.consoleOutput = fmt.Sprint(a...)
}

func TestRiceCookerProgram(t *testing.T) {
	logger := &testLogger{}
	consoleLog = logger

	t.Run("Initial State", func(t *testing.T) {
		riceCookerState = StateOff
		riceAndWaterAdded = false

		if riceCookerState != StateOff || riceAndWaterAdded {
			t.Errorf("Expected initial state: %s, riceAndWaterAdded: false", StateOff)
		}
	})

	t.Run("Display Initial State", func(t *testing.T) {
		riceCookerState = StateOff
		displayState()
		expectedOutput := fmt.Sprintf("\nCurrent state of the rice cooker: %s\n", StateOff)

		if logger.consoleOutput != expectedOutput {
			t.Errorf("Expected console output: %s, got: %s", expectedOutput, logger.consoleOutput)
		}
	})

	t.Run("Not Allow Cooking Without Adding Rice and Water", func(t *testing.T) {
		setRiceAndWaterAdded(false)
		plugIn()
		expectedOutput := "Error: Add rice and water before plugging in and starting cooking."

		if logger.consoleOutput != expectedOutput {
			t.Errorf("Expected console output: %s, got: %s", expectedOutput, logger.consoleOutput)
		}

		if riceAndWaterAdded {
			t.Error("Expected riceAndWaterAdded to be false")
		}
	})

	t.Run("Allow Adding Rice and Water", func(t *testing.T) {
		setRiceAndWaterAdded(false)
		setRiceAndWaterAdded(true)

		if !riceAndWaterAdded {
			t.Error("Expected riceAndWaterAdded to be true")
		}
	})

	t.Run("Display Error When Plugging In If Rice and Water Not Added", func(t *testing.T) {
		setRiceAndWaterAdded(false)
		plugIn()
		expectedOutput := "Error: Add rice and water before plugging in and starting cooking."

		if logger.consoleOutput != expectedOutput {
			t.Errorf("Expected console output: %s, got: %s", expectedOutput, logger.consoleOutput)
		}

		if riceCookerState != StateOff {
			t.Errorf("Expected riceCookerState to be %s, got: %s", StateOff, riceCookerState)
		}
	})

	t.Run("Display Error for End of Cooking Notification If Cooking is Not in Progress", func(t *testing.T) {
		setRiceAndWaterAdded(true)
		finishCooking()
		expectedOutput := "Error: No cooking in progress."

		if logger.consoleOutput != expectedOutput {
			t.Errorf("Expected console output: %s, got: %s", expectedOutput, logger.consoleOutput)
		}

		if riceCookerState != StateOff {
			t.Errorf("Expected riceCookerState to be %s, got: %s", StateOff, riceCookerState)
		}
	})

	t.Run("Quit the Program", func(t *testing.T) {
		quitProgram()
		expectedOutput1 := "\n======================================================================="
		expectedOutput2 := "Goodbye! Thanks for using the rice cooker program.\n"

		if logger.consoleOutput != expectedOutput1+expectedOutput2 {
			t.Errorf("Expected console output: %s%s, got: %s", expectedOutput1, expectedOutput2, logger.consoleOutput)
		}
	})
}
